### ISSUE TITLE
Non-release builds cause build to fail

### ISSUE THREAD

**Issue by theninthnode** [semantic-release/semantic-release#283]
I've set up semantic release for a module which has a CI pipeline in Codeship. If I do a commitizen style commit message for docs. i.e `docs('scope goes here')` sr throws an error that it couldn't find a new version for release. However this then causes the build to fail.

```
semantic-release ERR! pre Failed to determine new version.
semantic-release ERR! pre ENOCHANGE There are no relevant changes, so no new version is released.
```

Desired behavior would be to complete the build but WARN that a new version was not released. Otherwise I have to check that my tests didn't cause the failure.

I guess the issue is because of the following package.json script:

`"semantic-release": "semantic-release pre && npm publish && semantic-release post"`

In order to not run `npm publish` the `semantic-release pre` command must throw an error.

A possible fix would be to move those commands into a script to wrap that first command in a try catch to catch the ENOCHANGE errors and skip `npm publish` and `semantic-release post` so the build is allowed to pass?

**Comment by vrtxf** [comment]
check out https://www.npmjs.com/package/semantic-release#is-it-really-a-good-idea-to-release-on-every-push

**Comment by theninthnode** [comment]
My question is more around the CI build failing due to the error exception ENOCHANGE  when in fact it's just not being released. My tests are passing but SR aborts the build

**Comment by vrtxf** [comment]
I think SR was built for Travis with the actual `semantic-release` cli running on `travis_after_all`. It's basically a configuration issue, you are running `npm run semantic-release` inside your build when you should run it after everything is 'done'

**Comment by marvindanig** [comment]
@vrtxf  My `.travis.yml` file looks like this:

```
sudo: false
language: node_js
cache:
  directories:
    - node_modules
notifications:
  email: false
node_js:
  - '6'
  - '6.1'
  - '6.2'
  - '6.3'
   
before_install:
  - "npm i -g npm@latest"  
before_script:
  - npm prune
script:
  - npm run test:once
after_success:
  - npm run semantic-release
branches:
  except:
    - /^v\d+\.\d+\.\d+$


```

And the build fails exactly with the same error log as reported above. So what do you mean by "running npm run semantic-release inside your build when you should run it after everything is 'done'"?

The logs are here:

```

> flippy@0.0.0-semantically-released test:once /home/travis/build/marvindanig/flippy.JS
>  mocha ./tests/start.tests.js
  Flippy.JS
    First
      ✓ Should work
    new
      ✓ Is it installed correctly?
  2 passing (10ms)
The command "npm run test:once" exited with 0.
cache.2
store build cache
0.00s
1.89snothing changed, not updating cache
after_success
1.74s$ npm run semantic-release
> flippy@0.0.0-semantically-released semantic-release /home/travis/build/marvindanig/flippy.JS
> semantic-release pre && npm publish && semantic-release post
semantic-release ERR! pre Failed to determine new version.
semantic-release ERR! pre ENOCHANGE There are no relevant changes, so no new version is released.
npm ERR! Linux 3.13.0-40-generic
npm ERR! argv "/home/travis/.nvm/versions/node/v6.5.0/bin/node" "/home/travis/.nvm/versions/node/v6.5.0/bin/npm" "run" "semantic-release"
npm ERR! node v6.5.0
npm ERR! npm  v3.10.6
npm ERR! code ELIFECYCLE
npm ERR! flippy@0.0.0-semantically-released semantic-release: `semantic-release pre && npm publish && semantic-release post`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the flippy@0.0.0-semantically-released semantic-release script 'semantic-release pre && npm publish && semantic-release post'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the flippy package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     semantic-release pre && npm publish && semantic-release post
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs flippy
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls flippy
npm ERR! There is likely additional logging output above.
npm ERR! Please include the following file with any support request:
npm ERR!     /home/travis/build/marvindanig/flippy.JS/npm-debug.log
Done. Your build exited with 0.



```

**Comment by vrtxf** [comment]
@marvindanig OP said he is running it on Codeship CI, you show me Travis CI configuration for an entirely unrelated problem. As a side note, you have a different problem, did you read the documentation ? 
I can't find your package on NPM so it's unlikely I can be of any assistance

**Comment by marvindanig** [comment]
It's a new package on npm so not searchable yet: https://www.npmjs.com/package/flippy. I tried to publish a freshly mint npm package with just a `package.json` & `README` checked in using semver, commitizen and ghooks, but ended up publishing to npm manually because it simply wouldn't. And then tried it again with semver for the second pass, with a contrived `feat` commit, and it rolled out version 1.1.0 successfully and yet showed the error in the logs described above.

**Comment by vrtxf** [comment]
@marvindanig are you sure you read the documentation for semantic-release ? I ask this because the problems you are experiencing are the result of not reading the main page ... I don't want to be critical but it's explained clearly that SR is doing automated releases, in order to do this you need to have the package available and published already - at least, that's my understanding after using SR on 3 packages. My only issue with it was with last-release calculator that did not account for unpublished versions (for which I created a pull request and never got merged)

From the looks of it you might benefit from doing manual bumps, for example netflix unleash (still new technology and buggy in my opinion but might be less hassle for you)

**Comment by jhabdas** [comment]
@vrtxf we get your RTFM comments. As engineers, we should all give each other the benefit of the doubt and assume the docs were read before throwing out links expressing _opinion_ in the very same repo for which the users are having a problem:

The opinion is _emphasised_, emphasis mine:

> If you don’t feel comfortable making every passing feature or fix on your master branch addressable via npm you _might_ not treat your master right.

Pedantry aside, and I don't want to be critical, the problem expressed by @marvindanig is the _exact_ same problem OP described, which is:

```
semantic-release ERR! pre Failed to determine new version.
semantic-release ERR! pre ENOCHANGE There are no relevant changes, so no new version is released.

```

Now I have to ask, did you carefully read wither @theninthnode or @marvindanig's comments carefully? I ask, because I'm facing the exact same issue.

**Comment by jhabdas** [comment]
@theninthnode could you confirm if this is a dupe of https://github.com/semantic-release/semantic-release/issues/192 and close if it is? thanks!

**Comment by jhabdas** [comment]
@theninthnode Ignore my last comment. That other issue is a bunny hole IMHO, and yours is more clearly worded.

**Comment by vrtxf** [comment]
Yes, at the moment of writing I actually encountered that issue as well, found it as a quirk in the documentation, lead to a bug in last-release-npm that I fixed but never got accepted. 
I don't know what happened to that project as I only used it on 2 repos, some fixes which I found crucial never got accepted and I ended up being challenged due to the strict usage of SR.
I don't rate the project very highly based on my experience with it so if you want my advice I'd say use standard-version or unleash.
Cheers !

**Comment by jhabdas** [comment]
I was experiencing build failures with doc changes as well because I was running the SR pre script in the `script` step in Travis (my build depends on the version number created by it). As a workaround I chose to run my build twice. Once as a development dry run to fail the build if it doesn't work. And the second time after SR pre script, so I get the version number as a part of my build.

Here are the relevant parts of my Travis setup and [an example build](https://travis-ci.org/jhabdas/fetch-inject/builds/215236631):

```yaml
before_script:
  - npm prune
script:
  - npm run lint
  - npm run build
after_success:
  - npm run clean
  - npm run semantic-release:pre
  - npm run build
  - npm publish
  - npm run semantic-release:post
```

**Comment by pvdlg** [comment]
This issue should be solved with `semantic-release` version >=9.0.0 as we don't return an error code when there is no releases. See #480.

Please re-open if the issue is not solved.
